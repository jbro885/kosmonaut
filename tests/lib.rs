use crate::util::CommandUnderTest;
use std::ffi::OsStr;

mod util;

mod layout;
mod style;

pub fn reftest_expect_similar<S: AsRef<str> + AsRef<OsStr>>(file_path_one: S, file_path_two: S) {
    let (file_path_one, file_path_two): (&str, &str) =
        (file_path_one.as_ref(), file_path_two.as_ref());
    match run_similarity_cmd(file_path_one, file_path_two) {
        Ok(percent_similar) => {
            if percent_similar != 100.0f64 {
                panic!(format!("reftest_expect_similar failure.  files '{}' and '{}' were only {}% similar (expected 100%)", file_path_one, file_path_two, percent_similar));
            }
        }
        Err(err) => panic!(format!(
            "reftest_expect_similar *internal* failure (this is bad) with files '{}' and '{}'.  error message was: {}",
            file_path_one,
            file_path_two,
            err
        )),
    }
}

fn run_similarity_cmd<S: AsRef<str> + AsRef<OsStr>>(
    file_path_one: S,
    file_path_two: S,
) -> Result<f64, String> {
    let mut cmd = CommandUnderTest::new();
    cmd.arg("similarity");
    cmd.arg("--files");
    cmd.arg(file_path_one);
    cmd.arg(file_path_two);
    cmd.arg("--similarity-percent-only");
    cmd.succeeds();
    let stdout = cmd.stdout().trim();
    stdout.parse::<f64>().map_err(|_| stdout.trim().to_owned())
}
