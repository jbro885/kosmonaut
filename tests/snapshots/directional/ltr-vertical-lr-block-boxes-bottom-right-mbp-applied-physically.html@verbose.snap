---
source: tests/lib.rs
expression: dump_layout_cmd.stdout()
input_file: tests/websrc/directional/ltr-vertical-lr-block-boxes-bottom-right-mbp-applied-physically.html
---
HTML BlockContainer at (0, 0) size 1045x1080 (ml0 mr0 mb0 mt0 bl0 br0 bb0 bt0 pl0 pr0 pb0 pt0)
  BODY BlockContainer at (8, 8) size 1029x1064 (ml8 mr8 mb8 mt8 bl0 br0 bb0 bt0 pl0 pr0 pb0 pt0)
    DIV BlockContainer at (8, 8) size 300x200 (ml0 mr25 mb50 mt0 bl0 br12 bb25 bt0 pl0 pr6 pb12 pt0)
    DIV BlockContainer at (351, 8) size 300x200 (ml0 mr25 mb50 mt0 bl0 br12 bb25 bt0 pl0 pr6 pb12 pt0)
    DIV BlockContainer at (694, 8) size 300x200 (ml0 mr25 mb50 mt0 bl0 br12 bb25 bt0 pl0 pr6 pb12 pt0)

